### PR TITLE
feat: watsonx data module

### PR DESCRIPTION
**Pull request by iamar7**
### Description
Issue: Internal-[11374](https://github.ibm.com/GoldenEye/issues/issues/11374)

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [x] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: cra-config.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # More info about this file at https://github.com/terraform-ibm-modules/common-pipeline-assets/blob/main/.github/workflows/terraform-test-pipeline.md#cra-config-yaml
 version: "v1"
 CRA_TARGETS:
-  - CRA_TARGET: "examples/advanced" # Target directory for CRA scan. If not provided, the CRA Scan will not be run.
+  - CRA_TARGET: "examples/basic" # Target directory for CRA scan. If not provided, the CRA Scan will not be run.
     CRA_IGNORE_RULES_FILE: "cra-tf-validate-ignore-rules.json" # CRA Ignore file to use. If not provided, it checks the repo root directory for `cra-tf-validate-ignore-rules.json`
     PROFILE_ID: "0e6e7b5a-817d-4344-ab6f-e5d7a9c49520" # SCC profile ID (currently set to the FSCloud 1.4.0 profile).
```
Comment (daniel-butler-irl) [inline cra-config.yaml:6]: This would probably want to be changed to AI Security Guardrails 2.0 profile? @ocofaigh @vburckhardt

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -1,2 +1,3 @@
 # Primary owner should be listed first in list of global owners, followed by any secondary owners
-*       @SirSpidey @ocofaigh
+
+* @ocofaigh @daniel-butler-irl
```
Comment (daniel-butler-irl) [inline .github/CODEOWNERS:3]: I think you should add yourself and ask in the channel for a volunteer to be a secondary code owner. (We'll find one if there are no volunteers)

File: examples/basic/outputs.tf [hunk 3]
```diff
@@ -2,17 +2,27 @@
...
+  description = "The GUID of the watsonx.data instance."
+  value       = module.watsonx_data.watsonx_data_guid
+}
+
+output "watsonx_data_name" {
+  description = "The name of the watsonx.data instance."
+  value       = module.watsonx_data.watsonx_data_name
+}
+
+output "watsonx_data_plan_id" {
```
Comment (daniel-butler-irl) [inline examples/basic/outputs.tf:25]: since the module creates a resource group if one is not provided we should output the resource group name and id

File: examples/basic/variables.tf [hunk 4]
```diff
@@ -2,32 +2,35 @@
...
+variable "ibmcloud_api_key" {
+  description = "Used with the Terraform IBM-Cloud/ibm provider"
+  sensitive   = true
   type        = string
-  description = "Region to provision all resources created by this example"
-  default     = "us-south"
 }
 
-variable "prefix" {
+variable "location" {
```
Comment (daniel-butler-irl) [inline examples/basic/variables.tf:17]: this is the correct terminology for this service, but do we want to keep region for consistency across modules? @ocofaigh

File: .github/CODEOWNERS (hunk 2 above)
Comment (ocofaigh) [reply .github/CODEOWNERS:3]: +1

File: examples/basic/outputs.tf (hunk 3 above)
Comment (ocofaigh) [reply examples/basic/outputs.tf:25]: you mean the example creates the resource group, not the module

File: examples/basic/variables.tf (hunk 4 above)
Comment (ocofaigh) [reply examples/basic/variables.tf:17]: Agree - use `region`

File: cra-config.yaml (hunk 1 above)
Comment (iamar7) [reply cra-config.yaml:6]: @ocofaigh @vburckhardt Do we want to change this to AI Security Guardrails 2.0 profile?

Comment (iamar7) [reply cra-config.yaml]: As suggested in the scrum, I've updated the SCC profile id to AI Security Guardrails 2.0

File: README.md [hunk 5]
```diff
@@ -1,5 +1,4 @@
-<!-- Update this title with a descriptive name. Use sentence case. -->
-# Terraform modules template project
+# Terraform IBM Watsonx Data
 
 <!--
 Update status and "latest release" badges:
```
Comment (ocofaigh) [inline README.md:4]: Can you change the badge below from Incubating to Stable?

File: .github/settings.yml [hunk 6]
```diff
@@ -25,4 +25,4 @@ repository:
   # description: ""
 
   # Use a comma-separated list of topics to set on the repo (ensure not to use any caps in the topic string).
-  topics: terraform, ibm-cloud, terraform-module, core-team
+  topics: terraform, ibm-cloud, terraform-module, core-team, watsonx
```
Comment (ocofaigh) [inline .github/settings.yml:28]: maybe add `watsonx.data` ?

File: README.md [hunk 7]
```diff
@@ -1,5 +1,4 @@
-<!-- Update this title with a descriptive name. Use sentence case. -->
-# Terraform modules template project
+# Terraform IBM Watsonx Data
```
Comment (ocofaigh) [inline README.md:1]: The service name is actually watsonx.data so suggest to use that in the title

File: README.md [hunk 8]
```diff
@@ -20,15 +19,11 @@ For information, see "Module names and descriptions" at
 https://terraform-ibm-modules.github.io/documentation/#/implementation-guidelines?id=module-names-and-descriptions
 -->
 
-TODO: Replace this with a description of the modules in this repo.
```
Comment (ocofaigh) [inline README.md:23]: Can you add a description of the module here just like it says in the template? Maybe point to the watsonx.data official docs too?

File: README.md [hunk 9]
```diff
@@ -46,74 +41,70 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
-Use real values instead of "var.<var_name>" or other placeholder values
-unless real values don't help users know what to change.
--->
+### Usage
 
 ```hcl
+module "watsonx_data" {
+    source = "terraform-ibm-modules/watsonx-data/ibm"
+    version = "X.Y.Z" # Replace "X.X.X" with a release version to lock into a specific release
+    resource
```
Comment (ocofaigh) [inline README.md:54]: typo ?

File: README.md [hunk 10]
```diff
@@ -46,74 +41,70 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
+    watsonx_data_name = "watsonx-data"
+    service = "lakehouse"
+    region = "us-south"
+    plan = "lite"
+    resource_group_id     = "xxXXxxXXxXxXXXXxxXxxxXXXXxXXXXX"
+}
```
Comment (ocofaigh) [inline README.md:60]: can you fix the indentation so all the equal signs line up - just like our terraform fmt linter would do

File: README.md [hunk 11]
```diff
@@ -46,74 +41,70 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
-<!-- NO PERMISSIONS FOR MODULE
-If no permissions are required for the module, uncomment the following
-statement instead the previous block.
--->
-
-<!-- No permissions are needed to run this module.-->
-
+You need the following permissions to run this module.
 
+- Editor platform role on watsonx.data if you must provision.
```
Comment (ocofaigh) [inline README.md:68]: Can we keep the structure as it is in the module template please

File: variables.tf [hunk 12]
```diff
@@ -2,8 +2,42 @@
...
+  type        = string
+}
+
+variable "watsonx_data_plan" {
+  default     = "do not install"
+  description = "The plan that's used to provision the watsonx.data instance."
+  type        = string
+  validation {
+    condition = anytrue([
+      var.watsonx_data_plan == "do not install",
```
Comment (ocofaigh) [inline variables.tf:37]: I think we should support `do not install` here

File: variables.tf [hunk 13]
```diff
@@ -2,8 +2,42 @@
...
+}
+
+variable "resource_group_id" {
+  description = "The resource group ID where the Event Streams instance is created."
+  type        = string
+}
+
+variable "region" {
+  default     = "us-south"
+  description = "The region that's used with the IBM Cloud Terraform IBM provider. It's also used during resource creation."
```
Comment (ocofaigh) [inline variables.tf:17]: The descrption should include all supported values. Are we sure this list is valid?

File: variables.tf [hunk 14]
```diff
@@ -2,8 +2,42 @@
...
 ########################################################################################################################
 
-#variable "my_variable" {
-#  type        = string
-#  description = "A description of my variable"
-#  default     = "default_value"
-#}
+variable "watsonx_data_name" {
+  type        = string
+  description = "The name of the watsonx.data instance."
```
Comment (ocofaigh) [inline variables.tf:7]: This has no default value, meaning its a required input. However it would not be required if passing existing CRN right?

File: tests/pr_test.go [hunk 15]
```diff
@@ -2,30 +2,38 @@
...
-func setupOptions(t *testing.T, prefix string, dir string) *testhelper.TestOptions {
-	options := testhelper.TestOptionsDefaultWithVars(&testhelper.TestOptions{
-		Testing:       t,
-		TerraformDir:  dir,
-		Prefix:        prefix,
-		ResourceGroup: resourceGroup,
+// Current supported regions
+var validRegions = []string{
+	"us-south",
+	"eu-de",
```
Comment (ocofaigh) [inline tests/pr_test.go:17]: what about the other regions?

File: README.md [hunk 16]
```diff
@@ -46,74 +41,70 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
--->
+### Usage
 
 ```hcl
+module "watsonx_data" {
+    source = "terraform-ibm-modules/watsonx-data/ibm"
+    version = "X.Y.Z" # Replace "X.X.X" with a release version to lock into a specific release
+    resource
+    watsonx_data_name = "watsonx-data"
+    service = "lakehouse"
```
Comment (ocofaigh) [inline README.md:56]: this usage is wrong - there is no `service` input in the module

File: examples/basic/variables.tf [hunk 17]
```diff
@@ -2,32 +2,35 @@
...
+  default     = "watsonx"
 
-variable "resource_tags" {
-  type        = list(string)
-  description = "Optional list of tags to be added to created resources"
-  default     = []
+  validation {
+    condition     = var.prefix != "" && length(var.prefix) <= 25
+    error_message = "You must provide a value for the prefix variable and the prefix length can't exceed 25 characters."
+  }
```
Comment (ocofaigh) [inline examples/basic/variables.tf:35]: I wouldn't add validation to the example - just add it to the module, otherwise you need to maintain in 2 places

File: examples/basic/variables.tf [hunk 18]
```diff
@@ -2,32 +2,35 @@
...
 }
 
-variable "prefix" {
+variable "region" {
+  default     = "us-south"
+  description = "Used with the Terraform IBM-Cloud/ibm provider as well as resource creation."
   type        = string
-  description = "Prefix to append to all resources created by this example"
-  default     = "basic"
+  validation {
```
Comment (ocofaigh) [inline examples/basic/variables.tf:21]: I wouldn't add validation to the example - just add it to the module, otherwise you need to maintain in 2 places

File: examples/basic/variables.tf [hunk 19]
```diff
@@ -2,32 +2,35 @@
...
+  sensitive   = true
   type        = string
-  description = "Region to provision all resources created by this example"
-  default     = "us-south"
 }
 
-variable "prefix" {
+variable "region" {
+  default     = "us-south"
+  description = "Used with the Terraform IBM-Cloud/ibm provider as well as resource creation."
```
Comment (ocofaigh) [inline examples/basic/variables.tf:19]: can you use the description that came with the module template?

File: examples/basic/variables.tf [hunk 20]
```diff
@@ -2,32 +2,35 @@
...
   type        = string
-  description = "The IBM Cloud API Key"
-  sensitive   = true
+  description = "An existing resource group name to use for this example. If not specified, a new resource group is created."
+  default     = null
 }
 
-variable "region" {
+variable "ibmcloud_api_key" {
+  description = "Used with the Terraform IBM-Cloud/ibm provider"
```
Comment (ocofaigh) [inline examples/basic/variables.tf:12]: can you use the description that came with the module template?

File: .github/settings.yml (hunk 6 above)
Comment (ocofaigh) [reply .github/settings.yml:28]: oh and add `lakehouse` too

**Review by ocofaigh — CHANGES_REQUESTED** [review]
I left some comments.
Also:
- can you add a second example wich will test an existing instance - otherwise we have a test gap
- Can you bump the required terraform version to `>= 1.9.0` and use some cross variable validation in variables.tf ?

**Comment by maheshwarishikha** [comment]
Got one question - this module currently supports the provisioning of watsonx.data. Then why are we dealing with `existing_watsonx_data_instance_crn`? If user provides existing instance crn, then we are not doing anything extra other than outputting of service details. Sh we do that?

**Comment by ocofaigh** [comment]
As per discussions earlier proceed with supporting existing instance in module.
@iamar7 I think your missing access tag support. Can you implement (like [this](https://github.com/terraform-ibm-modules/terraform-ibm-key-protect/blob/e2ef103c660af2c753b9fd2629cc6c54ad94262c/main.tf#L62-L71)). This is one example that could be used to add access tag to an existing instance.

File: variables.tf (hunk 14 above)
Comment (iamar7) [reply variables.tf:7]: Added a default value.

**Comment by iamar7** [comment]
I need the base module code in the main branch to add it as source in the main.tf of `test/existing-resouce/` otherwise I would have to write the entire main branch code in the test directory.

File: variables.tf (hunk 12 above)
Comment (iamar7) [reply variables.tf:43]: I don't see any `do not install` plan in the [documentation](https://cloud.ibm.com/docs/watsonxdata?topic=watsonxdata-getting-started) so why are we adding this in the module?

Comment (ocofaigh) [reply variables.tf]: sorry I meant to say 
I think we should **not** support do not install here

File: README.md [hunk 21]
```diff
@@ -46,74 +42,75 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
-<!-- NO PERMISSIONS FOR MODULE
-If no permissions are required for the module, uncomment the following
-statement instead the previous block.
--->
-
-<!-- No permissions are needed to run this module.-->
+You need the following permissions to run this module.
 
+- Account Management service
+  - Administrator role
```
Comment (ocofaigh) [inline README.md:69]: 4 spaces for indentation

File: README.md [hunk 22]
```diff
@@ -46,74 +42,75 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
 
+- watsonx.data
+  - Editor platform
```
Comment (ocofaigh) [inline README.md:72]: 4 spaces for indentation

File: examples/basic/variables.tf [hunk 23]
```diff
@@ -10,24 +16,18 @@ variable "ibmcloud_api_key" {
...
-
-variable "resource_group" {
-  type        = string
-  description = "The name of an existing resource group to provision resources in to. If not set a new resource group will be created using the prefix variable"
-  default     = null
+  description = "The prefix to add to all resources created by this example."
+  default     = "watsonx"
 }
 
-variable "resource_tags" {
```
Comment (ocofaigh) [inline examples/basic/variables.tf:29]: why did you remove `resource_tags` from the example?

File: main.tf [hunk 24]
```diff
@@ -1,3 +1,61 @@
...
+}
+
+resource "ibm_resource_instance" "data_instance" {
+  count             = var.existing_watsonx_data_instance_crn != null ? 0 : 1
+  name              = var.watsonx_data_name
+  service           = "lakehouse"
+  plan              = var.watsonx_data_plan
+  location          = var.region
+  resource_group_id = var.resource_group_id
+
```
Comment (ocofaigh) [inline main.tf:37]: why are resource tags not here?

File: tests/existing-resources/main.tf [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+}
+module "resource_group" {
+  source  = "terraform-ibm-modules/resource-group/ibm"
+  version = "1.1.6"
+  # if an existing resource group is not set (null) create a new one using prefix
+  resource_group_name          = var.resource_group == null ? "${var.prefix}-resource-group" : null
+  existing_resource_group_name = var.resource_group
+}
+
+resource "ibm_resource_instance" "data_instance" {
```
Comment (ocofaigh) [inline tests/existing-resources/main.tf:19]: suggest to use the local module code to deploy the instance instead of provider

File: examples/existing-instance/main.tf [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+########################################################################################################################
+# Watsonx Data
+########################################################################################################################
+
+module "existing_watsonx_data_instance" {
+  source                             = "../../"
+  existing_watsonx_data_instance_crn = var.existing_watsonx_data_instance_crn
+  resource_group_id                  = module.resource_group.resource_group_id
```
Comment (ocofaigh) [inline examples/existing-instance/main.tf:20]: why would you pass `resource_group_id` here? Its only used when creating a new instance

File: variables.tf [hunk 27]
```diff
@@ -2,8 +2,60 @@
...
+  default     = null
+}
+
+variable "resource_group_id" {
+  type        = string
+  description = "The resource group ID where the watsonx data instance is created."
```
Comment (ocofaigh) [inline variables.tf:13]: `resource_group_id` is not required when passing existing instance, so it needs to have a default value of null. And you should add variable validation for this using the cross variable validation feature that came with tf 1.9

File: variables.tf [hunk 28]
```diff
@@ -2,8 +2,60 @@
...
 
-#variable "my_variable" {
-#  type        = string
-#  description = "A description of my variable"
-#  default     = "default_value"
-#}
+variable "watsonx_data_name" {
+  type        = string
+  description = "The name of the watsonx.data instance."
+  default     = null
```
Comment (ocofaigh) [inline variables.tf:8]: add variable validation to ensure this is not null is user is not passing existing instance

File: variables.tf [hunk 29]
```diff
@@ -2,8 +2,60 @@
...
+    error_message = "You must specify 'eu-de', 'eu-gb', 'jp-tok', 'au-syd' or 'us-south' as the IBM Cloud region."
+  }
+}
+
+variable "existing_watsonx_data_instance_crn" {
+  type        = string
+  description = "The CRN of the an existing watsonx.data instance. If no value is passed, and new instance will be provisioned."
+  default     = null
+
+  validation {
```
Comment (ocofaigh) [inline variables.tf:31]: would it not make more sense to have this validation in the `watsonx_data_name` variable?

File: variables.tf [hunk 30]
```diff
@@ -2,8 +2,60 @@
...
+    condition     = var.existing_watsonx_data_instance_crn == null ? length(var.watsonx_data_name) > 0 : true
+    error_message = "You must specify a value for 'watsonx_data_name' if 'existing_watsonx_data_instance_crn' is null."
+  }
+}
+
+variable "watsonx_data_plan" {
+  type        = string
+  description = "The plan that's used to provision the watsonx.data instance."
```
Comment (ocofaigh) [inline variables.tf:39]: include allowed values in the description

File: tests/pr_test.go [hunk 31]
```diff
@@ -2,30 +2,50 @@
...
+	"eu-gb",
+	"jp-tok",
+	"au-syd",
+}
+
+func setupOptionsBasicExample(t *testing.T, prefix string, dir string) *testhelper.TestOptions {
+	options := testhelper.TestOptionsDefault(&testhelper.TestOptions{
```
Comment (ocofaigh) [inline tests/pr_test.go:33]: why are you not using `TestOptionsDefaultWithVars` ? If you add resource_tags support you can use this function instead

File: tests/pr_test.go [hunk 32]
```diff
@@ -2,30 +2,50 @@
...
+		Testing:      t,
+		TerraformDir: dir,
 	})
+	terraformVars := map[string]interface{}{
```
Comment (ocofaigh) [inline tests/pr_test.go:37]: please test access tags. For approach, see https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/92d61a4bae845b0e960407d1e9070e2240f2dbfa/tests/pr_test.go#L60

**Review by ocofaigh — CHANGES_REQUESTED** [review]
see comments

File: variables.tf [hunk 33]
```diff
@@ -2,8 +2,72 @@
...
+}
+
+variable "existing_watsonx_data_instance_crn" {
+  type        = string
+  description = "The CRN of the an existing watsonx.data instance. If no value is passed, and new instance will be provisioned."
+  default     = null
+
+  validation {
+    condition     = var.existing_watsonx_data_instance_crn != null || var.resource_group_id != null
+    error_message = "When 'existing_instance_crn' is specified then 'resource_group_id' is not required."
```
Comment (ocofaigh) [inline variables.tf:45]: We don't need this kind of validation. Infact we probably don't wan't it either as it would break in a stack that supports passing existing instance.

File: variables.tf [hunk 34]
```diff
@@ -2,8 +2,72 @@
...
+  validation {
+    condition     = var.existing_watsonx_data_instance_crn == null ? length(var.watsonx_data_name) > 0 : true
+    error_message = "You must specify a value for 'watsonx_data_name' if 'existing_watsonx_data_instance_crn' is null."
+  }
+}
+
+variable "resource_group_id" {
+  type        = string
+  description = "The resource group ID where the watsonx data instance is created."
+  default     = null
```
Comment (ocofaigh) [inline variables.tf:19]: Validation should be added here to ensure this value is not `null` when `existing_watsonx_data_instance_crn` is `null`

File: variables.tf [hunk 35]
```diff
@@ -2,8 +2,72 @@
...
+
+variable "resource_tags" {
+  type        = list(string)
+  description = "Optional list of tags to be added to created resources"
+  default     = []
+}
+
+variable "region" {
+  type        = string
+  description = "The region to provision the watsonx data instance. "
```
Comment (ocofaigh) [inline variables.tf:30]: you have a trailing space at the end

**Review by ocofaigh — CHANGES_REQUESTED** [review]
final few comments

**Comment by iamar7** [comment]
> final few comments

I've resolved those comments

**Review by ocofaigh — APPROVED** [review]
